feat: add `TandemRequestProvider` for combined `RequestList` and `RequestQueue` usage

CodeMan62 opened the pull request:
## Overview
This PR introduces two new components to improve request management in Crawlee:

- `TandemRequestManager`: A provider that seamlessly combines `RequestList` and `RequestQueue`, allowing crawlers to use both sources efficiently.
- `RequestListAdapter`: An adapter that makes `RequestList` compatible with the `IRequestProvider` interface.

## Problem Solved
When using both `RequestList` and `RequestQueue` in a crawler, users often need to implement custom logic to ensure URLs aren't processed twice. This implementation standardizes and simplifies this pattern by:

1. First processing requests from the `RequestList`
2. Automatically transferring these requests to the `RequestQueue` in the background
3. Ensuring each URL is processed exactly once
4. Gracefully transitioning to `RequestQueue` after the list is exhausted

## Implementation Details
- The `RequestListAdapter` wraps a `RequestList` instance and adapts its interface to match `IRequestManager`.
- The `TandemRequestManager` orchestrates the flow between the list and queue:
  - It implements the `IRequestManager` interface
  - Handles request processing through the queue while ensuring list requests are enqueued first
  - Maintains proper request state tracking across both sources
  - Implements an on-demand transfer mechanism to move requests the same way as current BasicCrawler

## Integration
The implementation is already integrated into `BasicCrawler.initializeRequestManager()`, making it immediately available to all crawler types without requiring additional configuration changes.

## Backward Compatibility
This change is fully backward compatible:
- Existing code using either `RequestList` or `RequestQueue` alone will continue to work without changes
- This introduces an enhancement without modifying existing behavior

## Testing
The implementation was tested to ensure it correctly:
- Transfers requests from list to queue
- Processes requests in the correct order
- Handles errors appropriately
- Maintains proper request state

Closes #2499

barjin commented:
Please merge the latest `master` into your branch so we only get your changes in the diff. Cheers!

CodeMan62 commented:
@barjin Everything is done know you can review the PR thanks

barjin commented on packages/basic-crawler/src/internals/basic-crawler.ts:
```diff
@@ -1130,29 +1147,36 @@ export class BasicCrawler<Context extends CrawlingContext = BasicCrawlingContext
+            const requestListAdapter = new RequestListAdapter(this.requestList);
+            const tandemProvider = new TandemRequestProvider(requestListAdapter, this.requestQueue);
+            // Start the background transfer of requests
+            await tandemProvider.startBackgroundTransfer();
```
The `startBackgroundTransfer` method shouldn't be part of the public interface, since:
- it exposes internal implementation to the user
- it introduces [temporal coupling](https://en.wikipedia.org/wiki/Sequential_coupling). 
 
I'm not too sure about the background transfer in general (see my comment at `transferAllListRequestsToQueue`), maybe there's a better design choice we can take to solve both of those problems at once.

barjin commented on packages/basic-crawler/src/internals/basic-crawler.ts:
```diff
@@ -1072,7 +1086,10 @@ export class BasicCrawler<Context extends CrawlingContext = BasicCrawlingContext
+        await this._initializeRequestProviders();
         await this._loadHandledRequestCount();
+
+        this.events.on(EventType.MIGRATING, this._pauseOnMigration);
```
This event listener is attached later in the `run` implementation. 

https://github.com/apify/crawlee/blob/99af95e0dda511718b45cd41452589260c69909a/packages/basic-crawler/src/internals/basic-crawler.ts#L935

What is the reason to do it here?

barjin commented on packages/basic-crawler/src/internals/basic-crawler.ts:
```diff
@@ -1130,29 +1147,36 @@ export class BasicCrawler<Context extends CrawlingContext = BasicCrawlingContext
+        } else {
+            // Create and use a default RequestQueue if nothing else is provided
+            this.requestQueue = await this._getRequestQueue();
+            this.requestProvider = this.requestQueue;
```
We also need to access the right request source (`this.requestProvider`) over at 

https://github.com/apify/crawlee/blob/99af95e0dda511718b45cd41452589260c69909a/packages/basic-crawler/src/internals/basic-crawler.ts#L1325-L1326

barjin reviewed (changes requested):
Thank you for your initiative! 

There are some things I'm not too sure about (see my comments). Also, I see that your branch is almost a year (~500 commits) behind our master. Could you please merge the latest `apify/crawlee` master into your branch, so we can test the changes later? Cheers!

CodeMan62 commented:
> Thank you for your initiative!
> 
> There are some things I'm not too sure about (see my comments). Also, I see that your branch is almost a year (~500 commits) behind our master. Could you please merge the latest `apify/crawlee` master into your branch, so we can test the changes later? Cheers!

Hii @barjin  I am really too much confused about merging the latest master into my branch can you tell me how to do that please

barjin commented:
> > Thank you for your initiative!
> > There are some things I'm not too sure about (see my comments). Also, I see that your branch is almost a year (~500 commits) behind our master. Could you please merge the latest `apify/crawlee` master into your branch, so we can test the changes later? Cheers!
> 
> Hii @barjin I am really too much confused about merging the latest master into my branch can you tell me how to do that please

See the GitHub documentation at https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork?platform=linux . I see you might have already our `master` branch into your `master` branch - you want to sync our `master` into your `feat/tandem-request-provider`.

Try going [here](https://github.com/CodeMan62/crawlee/tree/feat/tandem-request-provider) and find the "Sync fork" button.

CodeMan62 commented:
It's done now let me resolve the requested changes Thanks for guide

CodeMan62 commented on packages/core/src/storages/tandem_request_provider.ts:
```diff
@@ -0,0 +1,133 @@
+     */
+    async fetchNextRequest<T extends Dictionary = Dictionary>(options?: RequestOptions): Promise<Request<T> | null> {
+        // Start the background transfer if not already started
+        if (!this.listFinishedPromise) {
```
So I have fixed it and also made some changes here take a look

CodeMan62 commented on packages/core/src/storages/tandem_request_provider.ts:
```diff
@@ -0,0 +1,133 @@
+     */
+    async isFinished(): Promise<boolean> {
+        const listFinished = await this.requestList.isFinished();
+        return listFinished && await this.requestQueue.isFinished();
```
Done thanks for preference :+1:

CodeMan62 commented:
Hey @barjin  Now the PR is ready to review/merge can you check and let me know if any changes required

barjin commented on packages/basic-crawler/src/internals/basic-crawler.ts:
```diff
@@ -1323,7 +1343,13 @@ export class BasicCrawler<Context extends CrawlingContext = BasicCrawlingContext
     protected async _runTaskFunction() {
-        const source = this.requestQueue || this.requestList || (await this.getRequestQueue());
+        // Ensure we have a request provider
+        if (!this.requestProvider) {
+            this.requestQueue = await this.getRequestQueue();
+            this.requestProvider = this.requestQueue;
+        }
```
Why not use the `initializeRequestProvider` function? By calling it, we ensure that `this.requestProvider` is defined, right?

barjin commented on packages/core/src/storages/tandem_request_provider.ts:
```diff
@@ -0,0 +1,133 @@
+     * Transfers all requests from the RequestList to the RequestQueue in the background.
+     * @private
+     */
+    private async transferAllListRequestsToQueue(): Promise<void> {
```
Also, does this `Promise` resolve only once `RequestList.isFinished()` is `true` for the original `requestList`? This might cause the Node process to hang "indefinitely" if the user decides to call something like `crawler.stop()`, right?

```typescript
const rl = RequestList([1000000000 requests]);

const crawler = new CheerioCrawler({ rl });
crawler.run();

sleep(5000);
await crawler.stop();
// The node process keeps running silently until all 1000000000 requests have been pushed to the RequestQueue
```

janbuchar commented:
Thanks @CodeMan62 for the contribution! I assume you're trying to fix https://github.com/apify/crawlee/issues/2499? If so, we should link it to this PR. 

I haven't reviewed the code yet, but here are some points I want to make:

1. Eventually, we want the tandem class to be called `RequestManagerTandem` so that it is consistent with the Python version of Crawlee. Can you rename `IRequestProvider` to `IRequestManager` as well? The `RequestProvider` and so on has to stay like that until we start working on a breaking release, so let's not rename that yet.

2.
> I still have some questions regarding the background transfer of the RequestList requests to the RequestQueue. I'm still not convinced this is the way to go here, especially since we've always done this lazily until now.

While issue https://github.com/apify/crawlee/issues/2499 mentions that we might want to transfer the request list independently of `fetchNextRequest` calls, I agree that we should tread carefully. Just extracting the tandem behavior as it is now should be sufficient. I made an initial draft a while back here https://github.com/apify/crawlee/pull/2898/files#diff-d631fbcd81c43142e84b9d2d01ead93451a95344001815f83c88f956ba14557a

CodeMan62 commented:
Hey @janbuchar Thanks for the feedback! I've made the following changes:
1. Renamed `TandemRequestProvider` to `RequestManagerTandem` to align with Python Crawlee's naming
2. Updated the implementation to use lazy loading with batched transfers (25 requests at a time) instead of background transfer
3. Left `IRequestProvider` as is for now, waiting for the breaking release to rename it to `IRequestManager`

And yes, this PR is addressing #2499 . I've looked at your initial draft and implemented a similar approach with batched transfers that should be more efficient while maintaining the same functionality.

barjin commented on packages/core/src/storages/tandem_request_provider.ts:
```diff
@@ -0,0 +1,142 @@
+        this.requestQueue = requestQueue;
+    }
+
+    /**
+     * Fetches the next request from the RequestQueue. If the queue is empty and the RequestList
+     * is not finished, it will transfer a batch of requests from the list to the queue.
+     * @private
+     */
```
Looks like a comment for `fetchNextRequest`?

barjin reviewed (commented):
I still have some comments, but they are not as pressing as the previous ones (and reflect rather my ideas than actual issues with the code).

Once we discuss the comments, I don't see any big issues anymore, so this PR is actually a soft go from me. Thanks!

CodeMan62 commented:
@barjin I have done the changes you have asked for take a look at the changes and let me know if any changes required

janbuchar commented:
> @janbuchar Regarding the rename to RequestManagerTandem - I've reverted this change for now. I think we should:
> 
> 1. Discuss the naming convention changes more broadly
> 2. Plan this as part of a coordinated update with other renames
> 3. Consider it for the next breaking release along with the IRequestProvider -> IRequestManager change
> 
> Let me know if you agree with this approach or if you'd prefer to handle the rename differently.

Hi @CodeMan62, i disagree. There is no good reason to introduce another set of names just to change them later. We should keep the current names backwards compatible for now, but where it's applicable, we should use the naming from the python port - that is the target state:

- instead of `TandemRequestProvider`, there should be `RequestManagerTandem`
- instead of `IRequestProvider`, we should have `IRequestManager`
- `RequestProvider` should implement `IRequestManager`

CodeMan62 commented:
Hey @janbuchar I have changed the names as you want can you please review it and tell me know if any code changes required

janbuchar reviewed (commented):
This is shaping up nicely!

- please rename the files to match the classes in them
- could you add some tests to verify the newly added functionality? 
- please run the code formatter (`yarn format`) to get rid of whitespace-only changes

CodeMan62 commented:
> This is shaping up nicely!
> 
> * please rename the files to match the classes in them
> * could you add some tests to verify the newly added functionality? Also, the existing tests must pass
> * please run the code formatter (`yarn format`) to get rid of whitespace-only changes

Working on it today, almost done with the commit

CodeMan62 commented:
Hey @janbuchar, I have renamed the files to match the classes in them and added tests, and all the tests are passing too, and also `yarn format` is done. Let me know if any changes are required! :)

janbuchar commented:
Hi @CodeMan62, I still see some test failures - could you look into that please?

CodeMan62 commented:
> Hi @CodeMan62, I still see some test failures - could you look into that please?

Yes, I am trying to fix them :)

CodeMan62 commented:
@janbuchar Can you  help me fix this failing tests please?

janbuchar commented:
@CodeMan62 I can try, but I'm stretched kinda thin right now. What did you already try? Is there anything you could tell me about the failing tests?

CodeMan62 commented:
> @CodeMan62 I can try, but I'm stretched kinda thin right now. What did you already try? Is there anything you could tell me about the failing tests?

almost 26 tests are failing let me try more to fix as much as I can and then let you know my mind is completely messed with these test but no worries let me put some more efforts

CodeMan62 commented:
@janbuchar i think i really need your help in it i am  not able to fix the tests help me in it please :pray:

barjin commented on test/core/request_manager_tandem.test.ts:
```diff
@@ -0,0 +1,239 @@
+        ]);
+        const requestQueue = await RequestQueue.open();
+
+        // Mock the addRequests function of requestQueue to verify it's called
+        const addRequestsSpy = vi.spyOn(requestQueue, 'addRequest');
```
```suggestion
        // Mock the addRequest function of requestQueue to verify it's called
        const addRequestSpy = vi.spyOn(requestQueue, 'addRequest');
```

barjin commented on packages/core/src/storages/request_manager_tandem.ts:
```diff
@@ -0,0 +1,162 @@
+     * @inheritdoc
+     */
+    async fetchNextRequest<T extends Dictionary = Dictionary>(): Promise<Request<T> | null> {
+        // If queue is empty, check if we can transfer more from list
+        const [listEmpty, listFinished] = await Promise.all([
+            this.requestList.isEmpty(),
+            this.requestList.isFinished(),
+        ]);
+
+        if (!listEmpty && !listFinished) {
+            await this.transferNextBatchToQueue();
+        }
```
the "_if queue is empty_" check is missing. This means that the transfer will happen on each `fetchNextRequest` call, which is likely not the desired behavior (judging by the comments).

barjin commented on packages/core/src/storages/request_manager_tandem.ts:
```diff
@@ -0,0 +1,162 @@
+     * Handles both successful transfers and failures appropriately.
+     * @private
+     */
+    private async transferNextBatchToQueue(): Promise<void> {
```
The `batch` is currently always of length 1, which doesn't sound too efficient.

janbuchar replied:
Yeah, I chose to do this to preserve the current behavior - I want to go one step at a time. This is just a private method and we can always make the batches bigger if we decide that it's a good idea.

barjin commented on packages/core/src/storages/request_manager_tandem.ts:
```diff
@@ -0,0 +1,162 @@
+    RequestsLike,
+} from './request_provider';
+
+/**
+ * A request manager that combines a RequestList and a RequestQueue.
+ * It first reads requests from the RequestList and then, when needed,
+ * transfers them in batches to the RequestQueue.
+ */
```
The description could be better. 

>  A request manager that ...

I believe "request manager" is a new term in this PR, right? Let's maybe stick to request provider, which (albeit not that popular), is already used in Crawlee? Also connected to:

> It first reads requests from the RequestList and then, when needed, transfers them in batches to the RequestQueue.

This sounds like transferring requests from one storage to another is the main use case for this class (it's providing requests to the crawler instance).

barjin commented on packages/core/src/storages/request_list_adapter.ts:
```diff
@@ -0,0 +1,102 @@
+     */
+    async reclaimRequest(
+        request: Request,
+        _options?: RequestQueueOperationOptions,
```
The `options` argument gets dropped silently here. Maybe log a warning `if(options)`? Or at least add a JSDoc comment about this?

janbuchar replied:
I wouldn't fret about this too much, it's an internal class that will be removed in v4.

barjin commented on packages/core/src/storages/request_provider.ts:
```diff
@@ -37,7 +37,68 @@ import { getRequestId, purgeDefaultStorages, QUERY_HEAD_MIN_LENGTH } from './uti
+    /**
+     * Returns number of handled requests.
+     */
+    handledCount(): Promise<number>;
+
+    /**
+     * Get the total number of requests known to the request manager.
+     */
+    getTotalCount(): number;
+
+    /**
+     * Get an offline approximation of the number of pending requests.
+     */
+    getPendingCount(): number;
```
nits:
- Not sure if making all those async wouldn't make this more future-proof. 
- Also, for the next breaking release, we might want to rename those to follow the same naming convention? (all `get` prefix - or all no prefix)?

janbuchar replied:
Yeah, in v4, I would like to do `IRequestList -> IRequestLoader`, then `IRequestManager extends IRequestLoader` and get rid of `RequestProvider`. Making mostly everything `async` is probably a good idea.

barjin reviewed (commented):
Thank you for pushing this!

I looked at this with a pair of fresh eyes and noticed some points I'd like to discuss:

janbuchar commented on test/core/request_manager_tandem.test.ts:
```diff
@@ -0,0 +1,239 @@
+        expect(reclaimSpy).toHaveBeenCalledWith(request, undefined);
+    });
+
+    test('handledCount returns the sum of list and queue handledCount', async () => {
+        const requestList = await RequestList.open(null, [
+            { url: 'https://example.com/1' },
+            { url: 'https://example.com/2' },
+        ]);
+        const requestQueue = await RequestQueue.open();
+
+        // Mock handledCount methods to return fixed values
+        vi.spyOn(requestList, 'handledCount').mockReturnValue(3);
+        vi.spyOn(requestQueue, 'handledCount').mockResolvedValue(2);
+
+        const tandem = new RequestManagerTandem(requestList, requestQueue);
+
+        // Only the request queue counts
+        await expect(tandem.handledCount()).resolves.toBe(2);
+    });
```
The name is wrong, thanks for pointing this out. When transferring requests from the request list to the request queue, we mark them as handled in the list, even though they are not *actually* handled.
